### ISSUE TITLE
Sina plot instead of beeswarm plot?

### ISSUE THREAD

**Issue by lrq3000** [ACCLAB/DABEST-python#58]
Hello,

I appreciate the approach taken by this framework a lot, and I would like to implement it in my publications. However, I would prefer to use a sina plot instead of a beeswarm, as it has 2 advantages:

1- apart from kernel density function estimation, it does not produce an artificial structuring on the data (ie, the "branch-like" lines in the beeswarm),

2- each class's sina plot's width is normalized across all classes, so that we can get an impression of the difference in sample size at a glance.

I think the last point in particular can very well complement the ideas put forward by the DABEST framework. There is a Python implementation of Sina plots in the plotnine package ([geom_sina](https://github.com/has2k1/plotnine/commit/358cea0cdf5c6c0f96c0b92a06a07c0e32b7fcad)).

Also maybe it would be interesting, if possible at all, to generalize the possibility of using other kinds of plots, as I guess different users might have different preferences?

**Comment by josesho** [comment]
Hi @lrq3000,

I agree about the superiority of sinaplots, especially when the Ns get very large. Our [R package](https://www.github.com/ACCLAB/dabestr) features sina plots; you could consider using that in lieu of this Python package.

I'm not familiar with the underlying implementation of `plotnine`; we use `matplotlib` and `seaborn` under the hood. There is an [implementation](https://github.com/mparker2/seaborn_sinaplot) of sinaplots in `seaborn` which should work. Feel free to submit a pull request if you come up with a working prototype!

> Also maybe it would be interesting, if possible at all, to generalize the possibility of using other kinds of plots, as I guess different users might have different preferences?

There are plenty of excellent general-purpose plotting packages for Python already. While we have plans to develop designs for other kinds of differences (e.g. differences in proportions), the DABEST suite will remain focussed on estimation plots: Gardner-Altman plots, and Cumming plots.

**Comment by lrq3000** [comment]
(working on it, thank you very much, didn't know about seaborn's implementation)

**Comment by josesho** [comment]
Closing it for now; feel free to reference this issue when doing your pull request!

**Comment by lrq3000** [comment]
About using other kinds of kde visualizations, for example if one has a LOT of samples, like 10000, per group, then a scatter-like plot becomes impractical, whereas a density plot such as a gradient plot or a violin plot would totally solve the issue. Furthermore, future visualizations may allow a better representation of the data, violin plots and sina plots are certainly not the end of it all. That's why I suggest to make the scatter-like plot generic, so that any kind of scatter-like plot can be plugged in. I'll see if this is possible, at worst I'll implement only the sina plot and violin plot.

Meanwhile, is it possible to reopen this issue please, to track the effort until it's done? Thank you :D

**Comment by josesho** [comment]
>  Furthermore, future visualizations may allow a better representation of the data, violin plots and sina plots are certainly not the end of it all. That's why I suggest to make the scatter-like plot generic, so that any kind of scatter-like plot can be plugged in. I'll see if this is possible, at worst I'll implement only the sina plot and violin plot.

Ah, thank you for clarifying.

Both the original [Gardner-Altman](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1339793) and [Cumming](https://www.routledge.com/Understanding-The-New-Statistics-Effect-Sizes-Confidence-Intervals-and/Cumming/p/book/9780415879682) designs implemented a swarmplot, and I'm strongly inclined to stick to the "display all data" paradigm. 

So I think a sina-plot (or some sort of force-directed layout, with dot size scaled to N size) is a definitely worthy effort. I'm not a fan of violin plots or box plots (despite their prevalence) because they can misrepresent differences between groups. 

![boxplots-violinplots](https://d2f99xq7vri1nk.cloudfront.net/BoxViolinSmaller.gif)
(Image taken from Fig 8 of this [page](https://www.autodeskresearch.com/publications/samestats).)

In summary, happy to accept a PR for sinaplots in DABEST-Python!

**Comment by mje-nz** [comment]
The Gardner-Altman figures use dot histograms, not swarm plots. In Cummings I can see violin plots (e.g. fig 6.4), strip plots (e.g. fig 6.6), and a whole lot of bare confidence intervals with no data (but I didn't look very hard).

I agree in principle that the data plots should show all the data, but swarm plots badly mis-represent the overall distribution for large samples. For example, compare these plots of two datasets:

![image](https://user-images.githubusercontent.com/6166539/65840432-96ab5380-e375-11e9-9e0d-3b1c47684a43.png)
![image](https://user-images.githubusercontent.com/6166539/65840433-9d39cb00-e375-11e9-8ca5-05ff466959c4.png)

For the small dataset I would argue only the box plot is really bad; most of the others give a reasonable impression of the distribution. For the large dataset I would argue the violin plot is the best, followed by the box plot and dot histogram; most of the others give a skewed impression of how big the tail is. The swarm plot is arguably the worst. You probably have different opinions, and that's fine! I expect a sina plot would do well on both datasets.

I think it would be reasonable to keep the default as a swarm plot (since sina plots seem hard) but let users pick a different data plot type. Would you accept a PR for that?

@lrq3000 are you still working on this?

**Comment by josesho** [comment]
Hi @mje-nz, 

My own inclination is to somehow drag-drop in @mparker2's seaborn-based [implementation](https://github.com/mparker2/seaborn_sinaplot). I suppose we could import the relevant `.py` file from that repo, and give the user an option between swarmplots and sinaplots.

Personally, I'm disinclined to use violin, box, and boxen plots to display rawdata. (Especially the violin plot, which we already use to display the bootstrap effect size.)

I should find some time to work on this in the next few weeks .... 🤞

If you can get @mparker2's code working within `dabest`, I'd be very happy to accept your PR!

**Comment by lrq3000** [comment]
Yes I am still working on it, first I need to solve #67 :-) Sorry I'm taking some time, it's not complicated, it's just that I spent more time reading statistical literature to get up to date with good practices. I have several other features I'd like to implement, but this one first :-) Help is of course welcome anyway!

**Comment by ValdarT** [comment]
Any developments here? The beeswarm is a problem even with only thousands of data points which is not that many: it is not that informative in this case and takes A LOT of time to plot currently.

**Comment by lrq3000** [comment]
@ValdarT Sorry I had unexpected events lately and had to stop working on it, but I still plan to, I'll try to finish it next week hopefully (as in any case I need it myself for my current main work).

**Comment by JAnns98** [comment]
Hi @lrq3000, It's been a long time since this thread was active! We do agree that swarmplot has its own set of issues when dealing with larger sample sizes and are hoping to work on making changes. Do you still have any work-in-progress code for this?

**Comment by lrq3000** [comment]
Unfortunately I don't have any progress to report on this feature and I don't have time to work on this in the foreseeable future.

17 juil. 2024 11:14:58 JAnns98 ***@***.***>:

**Comment by rhuszar** [comment]
I ran into the same obstacle to using this in practice.
It is really unfortunate the authors of the package don't address this issue themselves. The whole point of the DABEST paper was to provide a tool for the community to do better statistics, but plotting every datapoint is really impractical in the age of big data.
